exonum/src/storage/proof_list_index/mod.rs

```diff
@@ -59,17 +65,28 @@ pub struct ProofListIndexIter<'a, V> {
     base_iter: BaseIndexIter<'a, ProofListKey, V>,
 }
 
+/// TBD
+pub fn hash_leaf<V: StorageValue>(value: V) -> Hash {
```

Comment:
I wonder, can we modify `StorageValue` signature to avoid needless clone here?

Reply:
Like this?

```rust
pub trait StorageValue: UniqueHash + Sized {
    fn into_bytes(&self) -> Vec<u8>;

    fn from_bytes(value: &[u8]) -> Self;
}
```

As far as I understand, currently `into_bytes()` consumes `self` and `from_bytes()` takes `Cow<[u8]>` in order to use `std::String::into_bytes()` and `std::Vec::into_owned()` in `StorageValue` impls for `String` and `Vec`. Thus we avoid doing copies of possibly large strings and vectors. (Other types implementing `StorageValue` have fixed size so copying is at least predictable for them.)

However, with proposed changes it seems we will be doing clones of _all_ values so this optimization may be moot.

If this optimization is important then maybe we could _add_ a method to `StorageValue` to avoid breaking everything. For example, `fn as_bytes(&self) -> Cow<[u8]>` which will try avoiding expensive copies.

Reply:
I tried to avoid reimplementing `StorageValue` for all types. I thought about extending `UniqueHash` 
```
impl<T: CryptoHash + AsRef<[u8]> UniqueHash for T {
    fn hash(&self) -> Hash {
        CryptoHash::hash(self)
    }

    fn hash_to(&self, stream: HashStream) -> HashStream {
        stream.update(&self.as_ref())
    }
}
```
But in this case `T` is need to be bound with something like `AsRef`. And we also decided to get rid of `HashStream`.